### ISSUE TITLE
Fix CHAPI route nav on iOS

### ISSUE THREAD

**Issue by dmitrizagidulin** [digitalcredentials/learner-credential-wallet#341]
PR #326 enabled receiving CHAPI credential store requests to LCW. However, as Kayode mentioned in the PR notes, there's still a hitch in the workflow -- it only receives the credential on the second try. Meaning:

The first time you visit https://playground.chapi.io/issuer on the same device LCW is running, select the JFF credential, then click Generate, then click Store, then click 'Select Native App Instead' button on the bottom, select LCW, it opens LCW but gives the following error in the console:

```
[javascript] The navigation state parsed from the URL contains routes not present in the root navigator. 
This usually means that the linking configuration doesn't match the navigation structure. 
See https://reactnavigation.org/docs/configuring-links for more details on how to specify a linking configuration.
```
The second time (if you then switch over to the browser and click 'Select Native App Instead' button again), it goes into Available Credentials / receive VC screen, correctly.

@sethduffin - can you take a look at how the routing is set up, to see if you can fix this?

**Comment by sethduffin** [comment]
Works on Android, not iOS, ensure the route is opening on first launch.
_Time box this to 30min_

**Comment by sethduffin** [comment]
I'm having a hard time reproducing this issue on a simulator (the exchange is successfully recognized and the credential is added - even on first launch). I'll need to continue with different environments to try to repro (@dmitrizagidulin let me know if you're aware of any environmental factors that could be related)

If I had to theorize on the cause of the issue, I would suspect that the navigation intent is being handled before the store/database is initialized. This aligns with the error message of `the URL contains routes not present in the root navigator` because only the onboarding navigation route is available before runtime initialization (due to the nature of secure routing).

Having met the 30min time box, I'm holding off on trying to repro, but I'm happy to pick this up if/when the limit get's extended.

**Comment by dmitrizagidulin** [comment]
Thanks @sethduffin !

**Comment by kezike** [comment]
**Context:**
As I mentioned to @kayaelle on last week's TWG call, I found a fix for this iOS CHAPI bug in a different app a couple months ago, but never had a chance to address it, as it wasn’t a DCC priority at the time. I understand that there is a new CHAPI flow in development, but until that is integrated into the playground - and in order to provide continued support for the current workflow - I will add some notes here to explain the issue and the solution.

**Problem:**
The root of the issue is that the `react-native-receive-sharing-intent` library  is invoking event handlers (i.e., `Linking.getInitialURL` and `Linking.addEventListener`) that are already invoked from the main app and proceeding only after its own call to these libraries. Essentially, the CHAPI flow doesn’t proceed because the calls to these event handlers are not being detected when the `react-native-receive-sharing-intent` library invokes. Consequently, any logic that is in the event handler callback is not executed.

**Solution:**
In order to fix this issue, I created a patch to the `react-native-receive-sharing-intent` library that removes the calls to  `Linking.getInitialURL` and `Linking.addEventListener` and placed the content of the contained callbacks outside of the scope of these event handlers. Below, I have included a critical snapshot of the patch as well as an attachment to the full patch. As you can see, this patch comes with a slight interface change, which also needs to be reflected in the calling code.

_Patch snapshot:_
```
-  getReceivedFiles(handler, errorHandler, protocol = "ShareMedia") {
+  getReceivedFiles(handler, errorHandler, url) {
+    let protocol = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : "ShareMedia";
     if (this.isIos) {
-      Linking.getInitialURL().then(res => {
-        if (res && res.startsWith(`${protocol}://dataUrl`) && !this.isClear) {
-          this.getFileNames(handler, errorHandler, res);
-        }
-      }).catch(() => {});
-      Linking.addEventListener("url", res => {
-        const url = res ? res.url : "";
-
-        if (url.startsWith(`${protocol}://dataUrl`) && !this.isClear) {
-          this.getFileNames(handler, errorHandler, res.url);
-        }
-      });
+      if (url && url.startsWith(`${protocol}://dataUrl`) && !this.isClear) {
+        this.getFileNames(handler, errorHandler, url);
+      }
```

_Full patch:_
[react-native-receive-sharing-intent+2.0.0.patch](https://github.com/digitalcredentials/learner-credential-wallet/files/11478375/react-native-receive-sharing-intent%2B2.0.0.patch)

**Comment by kayaelle** [comment]
@dmitrizagidulin - please provide info on new CHAPI methods. 

Do you recommend that we fix the implementation we have currently and then creating a new issue for CHAPI methods? Or create two new issues for new iOS & android CHAPI implementation?

**Comment by kayaelle** [comment]
Can close. Replaced by: [https://github.com/digitalcredentials/learner-credential-wallet/issues/457](https://github.com/digitalcredentials/learner-credential-wallet/issues/457)
